core/src/main/java/tc/oc/pgm/stats/StatType.java

```diff
@@ -1,43 +1,81 @@
-import static net.kyori.adventure.text.format.NamedTextColor.*;
+import static net.kyori.adventure.text.format.NamedTextColor.GREEN;
+import static net.kyori.adventure.text.format.NamedTextColor.RED;
+import static net.kyori.adventure.text.format.NamedTextColor.YELLOW;
 import static tc.oc.pgm.stats.StatsMatchModule.damageComponent;
 import static tc.oc.pgm.util.text.NumberComponent.number;
 
 import java.util.Locale;
 import net.kyori.adventure.text.Component;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.util.math.Formula;
 
-public enum StatType {
-  KILLS,
-  DEATHS,
-  ASSISTS,
-  KILL_STREAK,
-  BEST_KILL_STREAK,
-  KILL_DEATH_RATIO,
-  LONGEST_BOW_SHOT {
-    private final String blocks = key + ".blocks";
+public sealed interface StatType<I> {
+  Component makeNumber(Number number);
+
+  Component buildMessage(Component valueComponent);
+
+  Component component(I input);
+
+  enum Builtin implements StatType<StatHolder> {
+    KILLS,
+    DEATHS,
+    ASSISTS,
+    KILL_STREAK,
+    BEST_KILL_STREAK,
+    KILL_DEATH_RATIO,
+    LONGEST_BOW_SHOT {
+      private final String blocks = key + ".blocks";
+
+      @Override
+      public Component makeNumber(Number number) {
+        return translatable(blocks, number(number, YELLOW));
+      }
+    },
+    DAMAGE {
+      @Override
+      public Component makeNumber(Number number) {
+        return damageComponent(number.doubleValue(), GREEN);
+      }
+    };
+
+    public final String key = "match.stats.type." + name().toLowerCase(Locale.ROOT);
 
     @Override
     public Component makeNumber(Number number) {
-      return translatable(blocks, number(number, YELLOW));
+      return number(number, this == DEATHS ? RED : GREEN);
     }
-  },
-  DAMAGE {
+
     @Override
-    public Component makeNumber(Number number) {
-      return damageComponent(number.doubleValue(), GREEN);
+    public Component component(StatHolder stats) {
+      return translatable(key, makeNumber(stats.getStat(this)));
     }
-  };
-
-  public final String key = "match.stats.type." + name().toLowerCase(Locale.ROOT);
-  public static final StatType[] VALUES = values();
 
-  public Component makeNumber(Number number) {
-    return number(number, this == DEATHS ? RED : GREEN);
+    @Override
+    public Component buildMessage(Component valueComponent) {
+      return translatable(key, valueComponent);
+    }
   }
 
-  public Component component(StatHolder stats) {
-    return translatable(key, makeNumber(stats.getStat(this)));
+  record OfFormula(Formula<MatchPlayer> formula, Component statDisplayName, double defaultValue)
+      implements StatType<MatchPlayer> {
+    private static final String KEY = "match.stats.type.generic";
+
+    @Override
+    public Component makeNumber(Number number) {
+      return number(number, GREEN);
+    }
+
+    @Override
+    public Component component(MatchPlayer player) {
+      return translatable(KEY, statDisplayName, makeNumber(formula.apply(player)));
```

Comment:
```suggestion
      return buildMessage(makeNumber(formula.apply(player)));
```
seems like both methods are doing the same, just one takes a component while the other takes the stat holder from which to create the component, maybe both should just `component`?

Reply:
Seems reasonable can do